In [123]:
import os
import sys
parent_dir = os.path.dirname(os.getcwd())
sys.path.append(parent_dir)
utils_dir = os.path.join(parent_dir, "src", "utils")
sys.path.append(utils_dir)
import pandas as pd
import plotly.graph_objects as go
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import GRU, Dense, Dropout
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau
from process_data import process_data
from weighted_accuracy import weighted_accuracy_scorer
from plot_learning_curves import plot_learning_curves
import xgboost as xgb
import seaborn as sns

In [124]:
X_path= os.path.join("..", "data","input", "X_train_Wwou3IE.csv")
X_preprocessed = pd.read_csv(X_path, delimiter=',')
y_path= os.path.join("..", "data","input", "y_train_jJtXgMX.csv")
y_preprocessed = pd.read_csv(y_path, delimiter=',')

In [125]:
cold_rate = {
    1: 100,  # January
    2: 90,   # February
    3: 70,   # March
    4: 50,   # April
    5: 30,   # May
    6: 10,   # June
    7: 0,    # July
    8: 5,    # August
    9: 20,   # September
    10: 40,  # October
    11: 60,  # November
    12: 80   # December
}

X_preprocessed['DELIVERY_START'] = pd.to_datetime(X_preprocessed['DELIVERY_START'], utc = True)

X_preprocessed['month'] = X_preprocessed['DELIVERY_START'].dt.month
X_preprocessed['cold_rate'] = X_preprocessed['month'].map(cold_rate)
X_preprocessed.drop('month', axis=1, inplace=True)

X_preprocessed

DELIVERY_START  load_forecast  coal_power_available  \
0     2022-01-01 01:00:00+00:00        49439.0                3386.0   
1     2022-01-01 02:00:00+00:00        46511.0                3386.0   
2     2022-01-01 03:00:00+00:00        45158.0                3386.0   
3     2022-01-01 04:00:00+00:00        44779.0                3386.0   
4     2022-01-01 05:00:00+00:00        45284.0                3386.0   
...                         ...            ...                   ...   
10600 2023-03-29 17:00:00+00:00        50814.0                3386.0   
10601 2023-03-29 18:00:00+00:00        50628.0                3386.0   
10602 2023-03-29 19:00:00+00:00        48201.0                3386.0   
10603 2023-03-29 20:00:00+00:00        47967.0                3386.0   
10604 2023-03-29 21:00:00+00:00        48444.0                3386.0   

       gas_power_available  nucelear_power_available  \
0                  11487.0                   44118.0   
1                  11487.0                   44118.0   
2                  11487.0                   44118.0   
3                  11487.0                   44118.0   
4                  11487.0                   44118.0   
...                    ...                       ...   
10600              11952.0                   38320.0   
10601              11952.0                   38320.0   
10602              11952.0                   38320.0   
10603              11952.0                   38320.0   
10604              11952.0                   38320.0   

       wind_power_forecasts_average  solar_power_forecasts_average  \
0                            3035.0                            0.0   
1                            3143.0                            0.0   
2                            3288.0                            0.0   
3                            3447.0                            0.0   
4                            3679.0                            0.0   
...                             ...                            ...   
10600                        7552.0                          651.0   
10601                        8338.0                          109.0   
10602                        9115.0                            0.0   
10603                        9636.0                            0.0   
10604                       10140.0                            0.0   

       wind_power_forecasts_std  solar_power_forecasts_std  \
0                     79.248348                   0.000000   
1                     61.776532                   0.000000   
2                     44.291112                   0.000000   
3                     36.127588                   0.000000   
4                     30.983023                   0.000000   
...                         ...                        ...   
10600                247.408490                   7.821622   
10601                155.795012                   2.534054   
10602                126.884684                   0.000000   
10603                156.669189                   0.000000   
10604                204.124773                   0.000000   

       predicted_spot_price  cold_rate  
0                       NaN        100  
1                       NaN        100  
2                       NaN        100  
3                       NaN        100  
4                       NaN        100  
...                     ...        ...  
10600                108.11         70  
10601                125.66         70  
10602                138.01         70  
10603                136.74         70  
10604                120.32         70  

[10605 rows x 11 columns]

In [126]:
X = process_data(X_preprocessed.copy(deep=True), "predicted_spot_price", None, "standard")
y = process_data(y_preprocessed.copy(deep=True), None, None, None)

In [127]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [128]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [129]:
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)
train_mse = mean_squared_error(y_train, y_train_pred)
test_mse = mean_squared_error(y_test, y_test_pred)

print(f"Training R^2: {train_r2:.2f}, Training MSE: {train_mse:.2f}")
print(f"Testing R^2: {test_r2:.2f}, Testing MSE: {test_mse:.2f}")

Training R^2: 0.00, Training MSE: 1554.19
Testing R^2: 0.01, Testing MSE: 998.38


In [130]:
#plot_learning_curves(model, X_train, y_train, X_test, y_test)

In [131]:
model = LinearRegression()
scores = cross_val_score(model, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert scores to positive to represent Mean Squared Error (MSE)
mse_scores = -scores

# Calculate the root mean squared error (RMSE) from MSE
rmse_scores = np.sqrt(mse_scores)

print("Scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Standard deviation:", rmse_scores.std())

Scores: [23.91247996 49.98194333 57.22693659 27.09167125 16.98709076]
Mean RMSE: 35.04002437915392
Standard deviation: 15.674354785500233


In [132]:
from sklearn.linear_model import Ridge
model = Ridge(alpha = 1000)
scores = cross_val_score(model, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert scores to positive to represent Mean Squared Error (MSE)
mse_scores = -scores

# Calculate the root mean squared error (RMSE) from MSE
rmse_scores = np.sqrt(mse_scores)

print("Scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Standard deviation:", rmse_scores.std())

Scores: [23.70459668 49.7431715  57.14180623 27.0065     16.94964291]
Mean RMSE: 34.90914346170718
Standard deviation: 15.651689902173006


In [133]:
from sklearn.preprocessing import PolynomialFeatures

poly_features = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly_features.fit_transform(X_train)
X_poly = poly_features.transform(X)
X_test_poly = poly_features.transform(X_test)
model = LinearRegression()
model.fit(X_train_poly, y_train)
y_train_pred = model.predict(X_train_poly)
y_test_pred = model.predict(X_test_poly)

train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)
train_mse = mean_squared_error(y_train, y_train_pred)
test_mse = mean_squared_error(y_test, y_test_pred)

print(f"Training R^2: {train_r2:.2f}, Training MSE: {train_mse:.2f}")
print(f"Testing R^2: {test_r2:.2f}, Testing MSE: {test_mse:.2f}")

Training R^2: 0.02, Training MSE: 1524.58
Testing R^2: 0.01, Testing MSE: 1002.14


In [134]:
#plot_learning_curves(model, X_train, y_train, X_test, y_test)

In [135]:
model = LinearRegression()
scores = cross_val_score(model, X_poly, y, cv=5, scoring='neg_mean_squared_error')

# Convert scores to positive to represent Mean Squared Error (MSE)
mse_scores = -scores

# Calculate the root mean squared error (RMSE) from MSE
rmse_scores = np.sqrt(mse_scores)

print("Scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Standard deviation:", rmse_scores.std())

Scores: [29.33080911 51.35939985 61.05343202 29.3070203  20.57925275]
Mean RMSE: 38.32598280545683
Standard deviation: 15.255216614675247


In [136]:
from sklearn.linear_model import Ridge
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Assuming X_train, y_train, X_test, and y_test are already defined

# Transforming the data to include polynomial features
poly_features = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly_features.fit_transform(X_train)
X_test_poly = poly_features.transform(X_test)

# Initializing the Ridge regression model with regularization strength alpha
# Note: You can adjust the alpha value to see how it affects the model's performance
alpha = 100000000  # This is a common starting point for regularization strength
ridge_model = Ridge(alpha=alpha)

# Fitting the model to the polynomial features
ridge_model.fit(X_train_poly, y_train)

# Predicting on both training and testing sets
y_train_pred = ridge_model.predict(X_train_poly)
y_test_pred = ridge_model.predict(X_test_poly)

# Calculating R^2 and Mean Squared Error (MSE) for both training and testing sets
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)
train_mse = mean_squared_error(y_train, y_train_pred)
test_mse = mean_squared_error(y_test, y_test_pred)

print(f"Training R^2: {train_r2:.2f}, Training MSE: {train_mse:.2f}")
print(f"Testing R^2: {test_r2:.2f}, Testing MSE: {test_mse:.2f}")

Training R^2: 0.00, Training MSE: 1560.98
Testing R^2: -0.00, Testing MSE: 1011.16


In [137]:
ridge_model = Ridge(alpha=alpha)
scores = cross_val_score(ridge_model, X_poly, y, cv=5, scoring='neg_mean_squared_error')

# Convert scores to positive to represent Mean Squared Error (MSE)
mse_scores = -scores

# Calculate the root mean squared error (RMSE) from MSE
rmse_scores = np.sqrt(mse_scores)

print("Scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Standard deviation:", rmse_scores.std())

Scores: [23.41873371 49.33468324 57.09980814 27.09802576 16.9203883 ]
Mean RMSE: 34.77432783160903
Standard deviation: 15.601726242580124


In [138]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Create a decision tree regressor model
tree_reg = DecisionTreeRegressor(max_depth=5)  # You can adjust the max_depth as needed

# Fit the model to the training data
tree_reg.fit(X_train, y_train)

# Predict on the training set and the test set
y_train_pred = tree_reg.predict(X_train)
y_test_pred = tree_reg.predict(X_test)

# Calculate the performance metrics
train_mse = mean_squared_error(y_train, y_train_pred)
test_mse = mean_squared_error(y_test, y_test_pred)
train_rmse = np.sqrt(train_mse)
test_rmse = np.sqrt(test_mse)

train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the performance metrics
print(f"Training RMSE: {train_rmse:.2f}, Training R^2: {train_r2:.2f}")
print(f"Testing RMSE: {test_rmse:.2f}, Testing R^2: {test_r2:.2f}")


Training RMSE: 38.19, Training R^2: 0.07
Testing RMSE: 30.86, Testing R^2: 0.06


In [139]:
tree_reg = DecisionTreeRegressor(max_depth=2)
scores = cross_val_score(tree_reg, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert scores to positive to represent Mean Squared Error (MSE)
mse_scores = -scores

# Calculate the root mean squared error (RMSE) from MSE
rmse_scores = np.sqrt(mse_scores)

print("Scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Standard deviation:", rmse_scores.std())

Scores: [23.42558606 49.70112953 57.06211571 27.09356241 16.90664129]
Mean RMSE: 34.83780700124896
Standard deviation: 15.6625481264992


In [140]:
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

"""# Generate a synthetic regression dataset
X, y = make_regression(n_samples=1000, n_features=20, noise=0.1, random_state=42)

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
"""

# Initialize the Gradient Boosting Regressor
gbr = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)

# Fit the model on the training data
gbr.fit(X_train, y_train)

# Predict on the testing data
y_pred = gbr.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R2): {r2:.2f}")


c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Mean Squared Error (MSE): 746.17
R-squared (R2): 0.26


In [141]:
from sklearn.datasets import make_regression
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import GradientBoostingRegressor
import numpy as np
"""
# Initialize the Gradient Boosting Regressor
gbr = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)

# Perform 5-fold cross-validation and compute the cross-validation scores
# Note: By default, cross_val_score uses R^2 as the score to evaluate. 
# For MSE, we need to specify 'neg_mean_squared_error' as the scoring parameter,
# and later convert it to positive MSE scores.
cv_scores = cross_val_score(gbr, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert negative MSE scores to positive values
mse_scores = -cv_scores

# Calculate the root mean squared error (RMSE) from MSE scores
rmse_scores = np.sqrt(mse_scores)

print("Cross-validation RMSE scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Standard deviation of RMSE:", rmse_scores.std())"""


'\n# Initialize the Gradient Boosting Regressor\ngbr = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)\n\n# Perform 5-fold cross-validation and compute the cross-validation scores\n# Note: By default, cross_val_score uses R^2 as the score to evaluate. \n# For MSE, we need to specify \'neg_mean_squared_error\' as the scoring parameter,\n# and later convert it to positive MSE scores.\ncv_scores = cross_val_score(gbr, X, y, cv=5, scoring=\'neg_mean_squared_error\')\n\n# Convert negative MSE scores to positive values\nmse_scores = -cv_scores\n\n# Calculate the root mean squared error (RMSE) from MSE scores\nrmse_scores = np.sqrt(mse_scores)\n\nprint("Cross-validation RMSE scores:", rmse_scores)\nprint("Mean RMSE:", rmse_scores.mean())\nprint("Standard deviation of RMSE:", rmse_scores.std())'

More complex model

In [142]:
from sklearn.datasets import make_regression
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import GradientBoostingRegressor
import numpy as np

"""# Initialize the Gradient Boosting Regressor with increased complexity
gbr_complex = GradientBoostingRegressor(n_estimators=200, learning_rate=0.1, max_depth=5, random_state=42)

# Perform 5-fold cross-validation and compute the cross-validation scores
cv_scores_complex = cross_val_score(gbr_complex, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert negative MSE scores to positive values
mse_scores_complex = -cv_scores_complex

# Calculate the root mean squared error (RMSE) from MSE scores
rmse_scores_complex = np.sqrt(mse_scores_complex)

print("Cross-validation RMSE scores:", rmse_scores_complex)
print("Mean RMSE:", rmse_scores_complex.mean())
print("Standard deviation of RMSE:", rmse_scores_complex.std())"""


'# Initialize the Gradient Boosting Regressor with increased complexity\ngbr_complex = GradientBoostingRegressor(n_estimators=200, learning_rate=0.1, max_depth=5, random_state=42)\n\n# Perform 5-fold cross-validation and compute the cross-validation scores\ncv_scores_complex = cross_val_score(gbr_complex, X, y, cv=5, scoring=\'neg_mean_squared_error\')\n\n# Convert negative MSE scores to positive values\nmse_scores_complex = -cv_scores_complex\n\n# Calculate the root mean squared error (RMSE) from MSE scores\nrmse_scores_complex = np.sqrt(mse_scores_complex)\n\nprint("Cross-validation RMSE scores:", rmse_scores_complex)\nprint("Mean RMSE:", rmse_scores_complex.mean())\nprint("Standard deviation of RMSE:", rmse_scores_complex.std())'

Easier model

In [143]:
from sklearn.datasets import make_regression
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import GradientBoostingRegressor
import numpy as np

# Initialize the Gradient Boosting Regressor with increased complexity
gbr_complex = GradientBoostingRegressor(n_estimators=50, learning_rate=0.1, max_depth=2, random_state=42)

# Perform 5-fold cross-validation and compute the cross-validation scores
cv_scores_complex = cross_val_score(gbr_complex, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert negative MSE scores to positive values
mse_scores_complex = -cv_scores_complex

# Calculate the root mean squared error (RMSE) from MSE scores
rmse_scores_complex = np.sqrt(mse_scores_complex)

print("Cross-validation RMSE scores:", rmse_scores_complex)
print("Mean RMSE:", rmse_scores_complex.mean())
print("Standard deviation of RMSE:", rmse_scores_complex.std())

c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemb

Cross-validation RMSE scores: [24.75300424 50.09622651 57.48569086 27.27478572 17.07339804]
Mean RMSE: 35.33662107489588
Standard deviation of RMSE: 15.61395975016201


From the more complex model and the easier model, we deduce if we were overfitting or not : We were overfitting

In [144]:
from sklearn.datasets import make_regression
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import GradientBoostingRegressor
import numpy as np

# Initialize the Gradient Boosting Regressor with increased complexity
gbr_complex = GradientBoostingRegressor(n_estimators=50, learning_rate=0.1, max_depth=2, random_state=42)

# Perform 5-fold cross-validation and compute the cross-validation scores
cv_scores_complex = cross_val_score(gbr_complex, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert negative MSE scores to positive values
mse_scores_complex = -cv_scores_complex

# Calculate the root mean squared error (RMSE) from MSE scores
rmse_scores_complex = np.sqrt(mse_scores_complex)

print("Cross-validation RMSE scores:", rmse_scores_complex)
print("Mean RMSE:", rmse_scores_complex.mean())
print("Standard deviation of RMSE:", rmse_scores_complex.std())

c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemb

Cross-validation RMSE scores: [24.75300424 50.09622651 57.48569086 27.27478572 17.07339804]
Mean RMSE: 35.33662107489588
Standard deviation of RMSE: 15.61395975016201


'n_estimators': [150, 250],  # Number of boosting stages to be run
'learning_rate': [0.15, 0.25],  # Learning rate shrinks the contribution of each tree
'max_depth': [5, 6],  # Maximum depth of the individual regression estimators
'min_samples_split': [1, 3],  # The minimum number of samples required to split an internal node
'min_samples_leaf': [2, 3],
>Best parameters found:  {'learning_rate': 0.15, 'max_depth': 6, 'min_samples_leaf': 2, 'min_samples_split': 3, 'n_estimators': 250}
Best cross-validation score (negative MSE):  -973.8543644050851
Mean Squared Error (MSE) with best parameters: 493.66
R-squared (R2) with best parameters: 0.51

    'n_estimators': [200, 250],  # Number of boosting stages to be run
    'learning_rate': [0.15, 0.2],  # Learning rate shrinks the contribution of each tree
    'max_depth': [5, 6],  # Maximum depth of the individual regression estimators
    'min_samples_split': [2, 3],  # The minimum number of samples required to split an internal node
    'min_samples_leaf': [2],
Fitting 5 folds for each of 16 candidates, totalling 80 fits
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
Best parameters found:  {'learning_rate': 0.2, 'max_depth': 6, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 250}
Best cross-validation score (negative MSE):  -971.1968402606602
Mean Squared Error (MSE) with best parameters: 509.69
R-squared (R2) with best parameters: 0.50

In [146]:
from sklearn.model_selection import GridSearchCV

# Parameter grid to search
param_grid = {
    'n_estimators': [200, 250],  # Number of boosting stages to be run
    'learning_rate': [0.15, 0.2],  # Learning rate shrinks the contribution of each tree
    'max_depth': [5, 6],  # Maximum depth of the individual regression estimators
    'min_samples_split': [2, 3],  # The minimum number of samples required to split an internal node
    'min_samples_leaf': [2],  # The minimum number of samples required to be at a leaf node
}
# > learning_rate': 0.2, 'max_depth': 5, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 200

# Initialize the Gradient Boosting Regressor
gbr = GradientBoostingRegressor(random_state=42)

# Initialize the Grid Search model
grid_search = GridSearchCV(estimator=gbr, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error', n_jobs=-1, verbose=2)

# Fit the Grid Search model
"""grid_search.fit(X_train, y_train)

# Print the best parameters and the corresponding score
print("Best parameters found: ", grid_search.best_params_)
print("Best cross-validation score (negative MSE): ", grid_search.best_score_)

# Predict on the testing data using the best found parameters
y_pred = grid_search.predict(X_test)

# Evaluate the model with the best parameters
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE) with best parameters: {mse:.2f}")
print(f"R-squared (R2) with best parameters: {r2:.2f}")"""

'grid_search.fit(X_train, y_train)\n\n# Print the best parameters and the corresponding score\nprint("Best parameters found: ", grid_search.best_params_)\nprint("Best cross-validation score (negative MSE): ", grid_search.best_score_)\n\n# Predict on the testing data using the best found parameters\ny_pred = grid_search.predict(X_test)\n\n# Evaluate the model with the best parameters\nmse = mean_squared_error(y_test, y_pred)\nr2 = r2_score(y_test, y_pred)\n\nprint(f"Mean Squared Error (MSE) with best parameters: {mse:.2f}")\nprint(f"R-squared (R2) with best parameters: {r2:.2f}")'

In [147]:
from sklearn.datasets import make_regression
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import GradientBoostingRegressor
import numpy as np

# Initialize the Gradient Boosting Regressor with increased complexity
# learning_rate': 0.2, 'max_depth': 6, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 250
gbr_complex = GradientBoostingRegressor(n_estimators=250, learning_rate=0.2, max_depth=6, random_state=42, min_samples_split = 2, min_samples_leaf = 2)

# Perform 5-fold cross-validation and compute the cross-validation scores
cv_scores_complex = cross_val_score(gbr_complex, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert negative MSE scores to positive values
mse_scores_complex = -cv_scores_complex

# Calculate the root mean squared error (RMSE) from MSE scores
rmse_scores_complex = np.sqrt(mse_scores_complex)

print("Cross-validation RMSE scores:", rmse_scores_complex)
print("Mean RMSE:", rmse_scores_complex.mean())
print("Standard deviation of RMSE:", rmse_scores_complex.std())

c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


KeyboardInterrupt: 

---

In [ ]:
from sklearn.datasets import make_regression
from sklearn.model_selection import cross_val_score
from sklearn.svm import SVR
import numpy as np

# Initialize the Support Vector Regressor
svr = SVR(kernel='linear')

# Perform 5-fold cross-validation and compute the cross-validation scores
# Note: By default, cross_val_score uses R^2 as the score to evaluate. 
# For MSE, we need to specify 'neg_mean_squared_error' as the scoring parameter,
# and later convert it to positive MSE scores.
cv_scores = cross_val_score(svr, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert negative MSE scores to positive values
mse_scores = -cv_scores

# Calculate the root mean squared error (RMSE) from MSE scores
rmse_scores = np.sqrt(mse_scores)

print("Cross-validation RMSE scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Standard deviation of RMSE:", rmse_scores.std())

c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages

Cross-validation RMSE scores: [23.63519043 49.65357997 57.065247   27.11044023 17.3584948 ]
Mean RMSE: 34.96459048593854
Standard deviation of RMSE: 15.519168904019235


In [148]:
reg = xgb.XGBRegressor(n_estimator = 1000)
reg.fit(X_train, y_train,
        eval_set = [(X_train, y_train), (X_test, y_test)],
        verbose = True)

[0]	validation_0-rmse:38.66813	validation_1-rmse:30.95246
[1]	validation_0-rmse:37.52104	validation_1-rmse:30.39418
[2]	validation_0-rmse:35.94488	validation_1-rmse:30.52696
[3]	validation_0-rmse:34.75663	validation_1-rmse:29.42517
[4]	validation_0-rmse:31.82708	validation_1-rmse:28.95460
[5]	validation_0-rmse:29.68750	validation_1-rmse:28.33683
[6]	validation_0-rmse:27.97336	validation_1-rmse:28.08091
[7]	validation_0-rmse:27.60334	validation_1-rmse:27.82393
[8]	validation_0-rmse:26.62542	validation_1-rmse:27.70515
[9]	validation_0-rmse:25.53083	validation_1-rmse:27.43615
[10]	validation_0-rmse:25.28164	validation_1-rmse:27.27962
[11]	validation_0-rmse:24.47325	validation_1-rmse:27.07921
[12]	validation_0-rmse:24.06108	validation_1-rmse:27.01435
[13]	validation_0-rmse:23.87718	validation_1-rmse:26.91239
[14]	validation_0-rmse:23.46667	validation_1-rmse:26.78130
[15]	validation_0-rmse:23.06399	validation_1-rmse:26.68756
[16]	validation_0-rmse:22.74354	validation_1-rmse:26.56820
[17]	va

c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\xgboost\core.py:160: UserWarning: [17:40:31] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0b3782d1791676daf-1\xgboost\xgboost-ci-windows\src\learner.cc:742: 
Parameters: { "n_estimator" } are not used.

  warnings.warn(smsg, UserWarning)


[19]	validation_0-rmse:21.49061	validation_1-rmse:26.07209
[20]	validation_0-rmse:21.18977	validation_1-rmse:25.97604
[21]	validation_0-rmse:21.05082	validation_1-rmse:25.98183
[22]	validation_0-rmse:20.94272	validation_1-rmse:25.82827
[23]	validation_0-rmse:20.40630	validation_1-rmse:25.56626
[24]	validation_0-rmse:20.13283	validation_1-rmse:25.47315
[25]	validation_0-rmse:20.05416	validation_1-rmse:25.45844
[26]	validation_0-rmse:19.58857	validation_1-rmse:25.23729
[27]	validation_0-rmse:19.44111	validation_1-rmse:25.17743
[28]	validation_0-rmse:19.24379	validation_1-rmse:25.07779
[29]	validation_0-rmse:18.87212	validation_1-rmse:25.12871
[30]	validation_0-rmse:18.41335	validation_1-rmse:25.03082
[31]	validation_0-rmse:18.16267	validation_1-rmse:24.88405
[32]	validation_0-rmse:17.87451	validation_1-rmse:24.72892
[33]	validation_0-rmse:17.64802	validation_1-rmse:24.64494
[34]	validation_0-rmse:17.50318	validation_1-rmse:24.52993
[35]	validation_0-rmse:17.35563	validation_1-rmse:24.465

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimator=1000, n_estimators=None,
             n_jobs=None, num_parallel_tree=None, ...)

In [149]:
y_pred = reg.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [150]:
mse, r2

(566.6401968003687, 0.43933926268869183)

In [151]:
reg = xgb.XGBRegressor(n_estimator = 1000)
"""reg.fit(X_train, y_train,
        eval_set = [(X_train, y_train), (X_test, y_test)],
        verbose = True)"""
cv_scores_complex = cross_val_score(gbr_complex, X, y, cv=5, scoring='neg_mean_squared_error')

# Convert negative MSE scores to positive values
mse_scores_complex = -cv_scores_complex

# Calculate the root mean squared error (RMSE) from MSE scores
rmse_scores_complex = np.sqrt(mse_scores_complex)

print("Cross-validation RMSE scores:", rmse_scores_complex)
print("Mean RMSE:", rmse_scores_complex.mean())
print("Standard deviation of RMSE:", rmse_scores_complex.std())

c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemble\_gb.py:424: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\emac\Documents\MachineLearningAvance\elmy\elmy\elmy_env\lib\site-packages\sklearn\ensemb

Cross-validation RMSE scores: [38.34562361 55.19911016 59.35552329 31.7381651  26.79500605]
Mean RMSE: 42.2866856414645
Standard deviation of RMSE: 12.844230749132231
